alertingest/ingester.py

```diff
@@ -0,0 +1,129 @@
+"""
+A worker which copies alerts and schemas into an object store backend.
+"""
+from typing import Tuple
+import io
+import ssl
+import struct
+import logging
+from dataclasses import dataclass
+from aiokafka import AIOKafkaConsumer, ConsumerRecord
+from alertingest.storage import AlertDatabaseBackend
+from alertingest.schema_registry import SchemaRegistryClient
+
+
+logger = logging.getLogger(__name__)
+logger.level = logging.DEBUG
+
+
+@dataclass
+class KafkaConnectionParams:
+    """
+    A bundle of data required to connect to Kafka.
+    """
+    host: str
+    topic: str
+    group: str
+    username: str
+    password: str
+
+
+class IngestWorker:
+    def __init__(
+        self,
+        kafka_params: KafkaConnectionParams,
+        backend: AlertDatabaseBackend,
+        registry: SchemaRegistryClient,
+    ):
+        """
+        Copies Rubin alert data from a Kafka broker to a database backend, using a
+        schema registry to make sense of the alert data.
+
+        All alert data is expected to be encoded in Confluent Wire Format.
+        """
+        self.kafka_params = kafka_params
+        self.backend = backend
+        self.schema_registry = registry
+
+    async def run(self, limit: int = -1, auto_offset_reset: str = "latest"):
+        """
+        Run the consumer, copying messages from Kafka to the IngestWorker's backend.
+
+        If limit > 0, exits after handling limit messages.
+        """
+        consumer = await self._create_consumer(auto_offset_reset)
+        await consumer.start()
+        try:
+            since_last_commit = 0
+            n = 0
+            logger.info("ingest worker run loop start")
+            async for msg in consumer:
+                logger.info("ingest worker received a message")
+                self.handle_kafka_message(msg)
+                logger.info("handle complete")
+                since_last_commit += 1
+                if since_last_commit == 100:
+                    logger.info("committing position in stream")
+                    await consumer.commit()
```

Comment:
I'm always interested in commits and restart strategies. So the approach here is that the worst-case is needing to re-do 100 messages, and the backend is such that's it's ok to try to re-write existing messages. neat :+1:

One thing you might do is make that 100 a configurable value.

Reply:
Right! Idempotence of the backend is a really powerful tool :)

It should definitely be configurable, I'll do that right now.